File: dann_jni/src/syncleus/dann/Synapse.java
package syncleus.dann;

public class Synapse
{
	static
	{
		System.load("/opt/sun-jre-bin-1.4.2.10/lib/i386/libdann.so");
	}
	
	
	private int nativeId = 0;
	private boolean constructed = false;
	
	
	public Synapse(Neuron Destination, Neuron Source, double InitialWeight)
	{
		this.nativeId = this.nativeConstructor(Destination, Source, InitialWeight);
		this.constructed = true;
	}
	
	protected int getNativeId()
	{
		return this.nativeId;
	}
	
	protected void finalize() throws Throwable
	{
		try
		{
			if( this.constructed == true )
			{
				this.nativeDestructor();
				this.nativeId = 0;
			}
		}
		finally
		{
			super.finalize();
		}
	}
	
	public native int nativeConstructor(Neuron Destination, Neuron Source, double InitialWeight);
	protected native void nativeDestructor();
	public native double getCurrentWeight();
	public native Neuron getDestinationNeuron();
	public native Neuron getSourceNeuron();
	public native double calculateOutput();
	public native void learnWeight();
	public native double calculateDifferential();
}

File: projects/c_dann_library/src/src/Synapse.cpp
#include "Synapse.h"
#include "Neuron.h"
#include "DNA.h"

Synapse::Synapse(Neuron* Destination, Neuron* Source, double InitialWeight)
{
	if((Destination == NULL) || (Source == NULL) || (InitialWeight == 0) )
		throw 0;

	this->DestinationNeuron = Destination;
	this->SourceNeuron = Source;
	this->Weight = InitialWeight;
}

Synapse::~Synapse()
{
}

double Synapse::GetCurrentWeight()
{
	return this->Weight;
}

Neuron* Synapse::GetDestinationNeuron()
{
	return this->DestinationNeuron;
}

Neuron* Synapse::GetSourceNeuron()
{
	return this->SourceNeuron;
}

double Synapse::CalculateOutput()
{
	return this->SourceNeuron->GetOutput() * this->Weight;
}

void Synapse::LearnWeight()
{
	this->Weight += this->DestinationNeuron->GetDNA()->LearningRate * this->SourceNeuron->GetOutput() * this->DestinationNeuron->GetDeltaTrain();
	if( this->Weight == 0 )
		this->DestinationNeuron->DisconnectSourceSynapse(this);
}

double Synapse::CalculateDifferential()
{
	return this->Weight * this->DestinationNeuron->GetDeltaTrain();
}

